h2oai/steam
# Proxy support

* Requires HAProxy 1.5+ to be already installed on the same machine Steam is running
* Generates HAproxy config on cluster launch/delete:
** binds on port 9999
** simple redirect based on URL path
** adds Authentication Basic header from a cookie (cookie name == cluster name)
* Generates a user/random password (username == steam username) for each new H2O cluster and starts H2O Jetty in secure mode (HashLogin) when "secure" checkbox in Steam is "on" - a base64 "user:password" token is stored in the DB (in the cluster table) 
* Adds a button to the Clusters screen for each cluster to go to them through the proxy (clicking on the name points to the clusters directly as it used to). Clicking it automatically adds a cookie (cookie name == cluster name) with the security token (from the DB) so people with the proxy URL won't be able to access it if they don't go through Steam first. 

How to test:

0) Make sure you can run `hadoop` locally (you either have a hadoop cluster set up on your machine or configured it to point to a remote one)
1) Install HAProxy 1.5+ (on mac simply `brew install haproxy`). Make sure to put it on your PATH.
2) Get this h2o-3 branch https://github.com/h2oai/h2o-3/tree/MD_context_path_doc
3) Run `./gradlew clean`
4) Get this h2o-flow branch https://github.com/h2oai/h2o-flow/tree/master
5) Run `make` in h2o-flow
6) In h2o-3 run `cp h2o-web/src/main/resources/www/flow/js/* h2o-web/lib/h2o-flow/build/js/`
7) In h2o3 build it with hadoop support https://github.com/h2oai/h2o-3/tree/MD_context_path_doc#6-building-h2o-on-hadoop
8) Pull this Steam branch, build it `go build`, clean the db `make db`
9) Run it with `sudo ./steam serve master --superuser-name=<USERNAME> --superuser-password=superuser`
10) Upload the engine you built in 6), launch a cluster (preferably with security checked)

Questions/TODOs:
1) I'm storing the token for H2O authentication in the DB, do we need a DB migration script?
2) The proxy button for each cluster is "reusing" the CSS for the trash button... Should we change that?
3) Currently using cluster name for URLs, IDs would be nicer (unique) but we get the cluster ID when its saved to the DB, which is after we start H2O - I need to pass that information to H2O so Jetty can set it as the context path, need to figure this part out.
4) We will need to add roles to cluster list so only certain people can see certain clusters.
